### PR TITLE
Suggested improvements

### LINKED ISSUE #5: Default routes for /robots.txt, /favicon.ico, /humans.txt

### PR DESCRIPTION
**Pull request by nottrobin**
- Add redirects for /favicon.ico, /robots.txt, /humans.txt
- Add errorhandlers for 404 and 500
- Remove config.py, which only had 1 config item
- Ask Talisker for "raven", "prometheus", "flask" by default - for our logging / error handling etc.

See https://github.com/canonical-web-and-design/canonical.com/pull/10/files#diff-bed92ddcb62003ca6b223dfa86df23b3L11 for example of how this would be used.

Fixes #5

QA
--

Pr for testing this here: https://github.com/canonical-web-and-design/canonical.com/pull/10
Demo here: https://canonical-com-canonical-web-and-design-pr-10.run.demo.haus

- [ ] https://canonical-com-canonical-web-and-design-pr-10.run.demo.haus/ works
- [ ] https://canonical-com-canonical-web-and-design-pr-10.run.demo.haus/test works
- [ ] https://canonical-com-canonical-web-and-design-pr-10.run.demo.haus/missing -> "404"
- [ ] https://canonical-com-canonical-web-and-design-pr-10.run.demo.haus/error -> "500"
- [ ] https://canonical-com-canonical-web-and-design-pr-10.run.demo.haus/favicon.ico -> Redirect to assets
- [ ] https://canonical-com-canonical-web-and-design-pr-10.run.demo.haus/robots.txt -> Redirect to `/static/robots.txt`
- [ ] https://canonical-com-canonical-web-and-design-pr-10.run.demo.haus/humans.txt -> Redirect to `/static/humans.txt`
- [ ] https://canonical-com-canonical-web-and-design-pr-10.run.demo.haus/hello -> 302 redirect to `/world` (404)
- [ ] https://canonical-com-canonical-web-and-design-pr-10.run.demo.haus/fish -> 301 redirect to `/chips` (404)

### CODE REVIEW

**Comment by jpmartinspt** [comment]
Have you considered http://flask.pocoo.org/docs/1.0/api/#flask.send_from_directory to server the static files instead of the redirect?

File: canonicalwebteam/flask_base/app.py [hunk 1]
```diff
@@ -17,32 +21,64 @@ def __init__(
         self,
         name,
         service,
-        prepare_deleted=prepare_deleted(),
```
Comment (jpmartinspt) [inline canonicalwebteam/flask_base/app.py]: By removing `prepare_deleted` and `prepare_redirects` we remove the possibility of customizing the deleted callback and to set/unset permanent redirects.

File: pyproject.toml [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
+readme = "README.md"
 
 [tool.poetry.dependencies]
-python = "^3"
+python = "^3.6"
 "canonicalwebteam.yaml-responses" = {version = "^1.1", extras = ["flask"]}
 flask = "1.0.2"
-talisker = {version = "0.14.3", extras = ["gunicorn"]}
+talisker = {version = "0.14.3", extras = ["gunicorn", "flask", "prometheus", "raven"]}
+poetry-setup = "^0.3.6"
```
Comment (jpmartinspt) [inline pyproject.toml]: Does this need to be set in dependencies and dev dependencies?

File: README.md [hunk 3]
```diff
@@ -1,27 +1,72 @@
 # Canonical Webteam Flask-Base
```
Comment (jpmartinspt) [inline README.md]: Could you add a little note on how to use poetry-setup in here?

File: pyproject.toml (hunk 2 above)
Comment (nottrobin) [reply pyproject.toml]: No, I'll remove that

File: README.md (hunk 3 above)
Comment (nottrobin) [reply README.md]: Yep will do

File: canonicalwebteam/flask_base/app.py (hunk 1 above)
Comment (nottrobin) [reply canonicalwebteam/flask_base/app.py]: This was very deliberate - I'd rather that each of the sites use the exact same interface for processing redirects and deleted pages - as in, they keep them in `redirects.yam` and `deleted.yaml` respectively. Allowing the function to be customised encourages sites to customise it.

I'd be more than happy to look for a `permanent-redirects.yaml` in this default configuration and parse permanent redirects from there, but I think it's better not to offer the ability for sites to customise their implementation.

**Comment by nottrobin** [comment]
> @jpmartinspt Have you considered http://flask.pocoo.org/docs/1.0/api/#flask.send_from_directory to server the static files instead of the redirect?

Yes I considered that. I thought it might be rather confusing if `/robots.txt` directly serves a file at `/static/robots.txt`, which would behave differently from all other `/static` files, and I did check that Google's bot at least will follow the redirect.

However, I'm open to the idea of asking for a file path instead of a URL, and simply loading the contents of that file. Do you think that would be better?

File: canonicalwebteam/flask_base/app.py (hunk 1 above)
Comment (jpmartinspt) [reply canonicalwebteam/flask_base/app.py]: I see your point, but with them being optional I don't think we are encouraging overriding. In any case the point was more about will we need permanent redirects or customized callbacks? I would be happy with a `permanent-redirects.yaml` or even tweaking `yaml-responses` so the status code could be set in `redirects.yaml`? Also happy to merge this as is and address this in a different PR.

**Comment by jpmartinspt** [comment]
> > @jpmartinspt Have you considered http://flask.pocoo.org/docs/1.0/api/#flask.send_from_directory to server the static files instead of the redirect?
> 
> Yes I considered that. I thought it might be rather confusing if `/robots.txt` directly serves a file at `/static/robots.txt`, which would behave differently from all other `/static` files, and I did check that Google's bot at least will follow the redirect.
> 
> However, I'm open to the idea of asking for a file path instead of a URL, and simply loading the contents of that file. Do you think that would be better?
> 
> In the case of `/favicon.ico` I definitely want to use a URL, because the actual favicon will likely live on the assets server. But I don't suppose that would be the case for `robots.txt` or `humans.txt` - that would probably be rather confusing.

I was thinking just serving them from the static folder like:

```
@app.route('/robots.txt')
@app.route('/sitemap.xml')
def static_from_root():
    return send_from_directory(app.static_folder, request.path[1:])
```

For the favicon:
Could we not just link the the asset server in the templates directly?

**Comment by nottrobin** [comment]
> I was thinking just serving them from the static folder like:
> 
> ```
> @app.route('/robots.txt')
> @app.route('/sitemap.xml')
> def static_from_root():
>     return send_from_directory(app.static_folder, request.path[1:])
> ```

Yeah that seems fine to me - I'd just wrap them in an `if os.path.isfile('robots.txt')`. So I'll remove the `robots_url` and `humans_url` options, and then document in the `README.md` that you should create those files at the root of the project if you want them served.

> For the favicon:
> Could we not just link the the asset server in the templates directly?

We do that already - e.g. https://ubuntu.com has `<link rel="icon" type="image/png" href="https://assets.ubuntu.com/v1/49a1a858-favicon-32x32.png" sizes="32x32" />`, but the issue here is that most browsers query `/favicon.ico` anyway, which is one of our highest 404s. So the idea here is to send something back at that path specifically. I think a redirect is the right solution in this case.

File: canonicalwebteam/flask_base/app.py [hunk 4]
```diff
@@ -17,32 +21,71 @@ def __init__(
...
 
-        self.before_request(prepare_redirects)
-        self.before_request(prepare_deleted)
+        if self.debug:
+            self.wsgi_app = DebuggedApplication(self.wsgi_app)
+
+        self.before_request(prepare_redirects())
+        self.before_request(
+            prepare_redirects(
+                path=os.path.join(self.root_path, "redirects.yaml"),
```
Comment (jpmartinspt) [inline canonicalwebteam/flask_base/app.py]: Should path be "permanent_redirects.yaml" here?

File: canonicalwebteam/flask_base/app.py [hunk 5]
```diff
@@ -17,32 +21,71 @@ def __init__(
...
+                return flask.redirect(favicon_url)
+
+        robots_path = os.path.join(self.root_path, 'robots.txt')
+        humans_path = os.path.join(self.root_path, 'humans.txt')
+
+        if os.path.isfile(robots_path):
+
+            @self.route("/robots.txt")
+            def robots():
+                flask.send_file(robots_path)
```
Comment (jpmartinspt) [inline canonicalwebteam/flask_base/app.py]: I believe this needs to be `return flask.send_file(robots_path)`.

File: canonicalwebteam/flask_base/app.py [hunk 6]
```diff
@@ -17,32 +21,71 @@ def __init__(
...
+
+        if os.path.isfile(humans_path):
+
+            @self.route("/humans.txt")
+            def robots():
+                flask.send_file(humans_path)
```
Comment (jpmartinspt) [inline canonicalwebteam/flask_base/app.py]: I believe this needs to be return flask.send_file(humans_path).

File: canonicalwebteam/flask_base/app.py (hunk 4 above)
Comment (nottrobin) [reply canonicalwebteam/flask_base/app.py]: absolutely

File: canonicalwebteam/flask_base/app.py (hunk 5 above)
Comment (nottrobin) [reply canonicalwebteam/flask_base/app.py]: Yep, I shouldn't have pushed so early =)

File: canonicalwebteam/flask_base/app.py (hunk 1 above)
Comment (nottrobin) [reply canonicalwebteam/flask_base/app.py]: I've made it look for `permanent-redirects.yaml` as well. Is that okay?

File: pyproject.toml [hunk 7]
```diff
@@ -1,18 +1,21 @@
...
 description = ""
 authors = ["Canonical webteam <webteam@canonical.com>"]
 packages = [{ include = "canonicalwebteam" }]
+readme = "README.md"
 
 [tool.poetry.dependencies]
-python = "^3"
+python = "^3.6"
 "canonicalwebteam.yaml-responses" = {version = "^1.1", extras = ["flask"]}
 flask = "1.0.2"
```
Comment (jpmartinspt) [inline pyproject.toml]: @tbille just noticed this should not be pinned, can we make it `>=`?

**Pull request merged**